On the Way to CDI 4.0 post. (#45)
cdi-spec/cdi-spec.org: news/2021-10-21-On_the_way_to_CDI_4.adoc

=== What is and isn’t Lite

The specification was split into Lite and Full parts in a way that Lite defines a set of core features and Full then builds on top of it, expanding sets of rules and presenting additional features. This means that CDI Lite is simply a subset of CDI Full - you can use Lite APIs in a portable way in CDI Full environments and every CDI Full implementation has to also implement CDI Lite. That being said, CDI Full remains ‘runtime oriented’ as opposed to Lite implementations which are likely to focus on build-time environments even though the specification sets no limits in this regard.

Below is a list of CDI features that are now exclusive to CDI Full implementations:

* Decorators
* Interception bound via `@Interceptors` annotation
** Interceptors bound via interceptor binding annotations work perfectly well in CDI Lite
* Specialization
* Passivation
* Conversation Scope

> * All other built-in scoped are present in CDI Lite

Reviewer: ```suggestion
** All other built-in scopes are present in CDI Lite
```

* Portable Extensions
** These are replaced by a new extension API which is described in the following paragraph

It is important to reiterate that existing applications are all considered to use CDI Full and as such will continue working unimpeded.

=== Introduction of Build Compatible Extensions

Portable Extensions are crucial for CDI extensibility but are also an ill fit for any build oriented framework. Therefore, an alternative extension SPI was proposed. This was covered in depth in a http://www.cdi-spec.org/news/2020/09/15/CDI_Lite_extension/[previous blog post] and has undergone many iterations since. However, its purpose stays ultimately the same - to provide an alternative for Portable Extensions that will work in all environments.

=== New Underlying Metamodel

> One of the goals of CDI Lite was to shake off as much reflection as possible. Nonetheless, current metamodel (`AnnotatedType` and friends) is very reflection-heavy which is why CDI Lite comes with new metamodel abstraction. Note that CDI Full can and will be able to use the old metamodel whereas CDI Lite, namely its extensions, will have to use the new metamodel.

Reviewer: ```suggestion
One of the goals of CDI Lite was to shake off as much reflection as possible. Nonetheless, current metamodel (`AnnotatedType` and friends) is very reflection-heavy which is why CDI Lite comes with new metamodel abstraction. Note that CDI Full and Portable Extensions will still use the old metamodel whereas CDI Lite and Build Compatible Extensions will use the new metamodel.
```

Maybe more precise?

This metamodel is currently housed inside CDI JAR but there has been some interest in using it from other Jakarta projects so it might end up being shipped as a separate artifact.

=== Shipping CDI 4.0 JAR

One of the questions on many meetings was around how to ship CDI JAR now that there is Lite and Full combined. The answer to that is that there is no clear way of achieving that without breaking existing applications through package name changes. Therefore, as a compromise, CDI will stay a single JAR containing all the APIs.

== Change of Default Empty `beans.xml` Discovery Mode

> One notable change coming to CDI is the https://github.com/eclipse-ee4j/cdi/issues/500[discovery mode of bean archives containing empty `beans.xml`]. These were formerly treated as explicit bean archives (e.g. having discovery mode `all`) but beginning with CDI 4.0, they will be considered as implicit bean archives (discovery mode `annotated`).

Reviewer: ```suggestion
One notable change coming to CDI is the https://github.com/eclipse-ee4j/cdi/issues/500[discovery mode of bean archives containing empty `beans.xml`]. These were formerly treated as explicit bean archives (i.e. having discovery mode `all`) but beginning with CDI 4.0, they will be considered as implicit bean archives (discovery mode `annotated`).
```

- i.e. = that is
- e.g. = for example
